include/exadg/convection_diffusion/driver.cpp

```diff
@@ -143,28 +144,124 @@ Driver<dim, Number>::ale_update() const
   time_int_bdf->ale_update();
 }
 
+template<int dim, typename Number>
+void
+Driver<dim, Number>::mark_cells_coarsening_and_refinement(dealii::Triangulation<dim> & tria,
+                                                          VectorType const & solution) const
+{
+  mark_cells_kelly_error_estimator(tria,
+                                   pde_operator->get_dof_handler(),
+                                   pde_operator->get_constraints(),
+                                   *pde_operator->get_mapping(),
+                                   solution,
+                                   application->get_parameters().degree +
+                                     1 /* n_face_quadrature_points */,
+                                   application->get_parameters().amr_data);
+}
+
+template<int dim, typename Number>
+void
+Driver<dim, Number>::setup_after_coarsening_and_refinement()
+{
+  // Update mapping
+  AssertThrow(ale_mapping.get() == 0,
+              dealii::ExcMessage(
+                "Combination of adaptive mesh refinement and ALE not implemented."));
+
+  std::shared_ptr<dealii::MappingQCache<dim>> mapping_q_cache =
+    std::dynamic_pointer_cast<dealii::MappingQCache<dim>>(mapping);
+  AssertThrow(
+    mapping_q_cache.get() == 0,
+    dealii::ExcMessage(
+      "Combination of adaptive mesh refinement and dealii::MappingQCache not implemented."));
+
+  pde_operator->setup_after_coarsening_and_refinement();
+
+  postprocessor->setup(*pde_operator);
```

Comment:
I think this can be problematic. Often, postprocessors have some history (e.g. collect statistics over time).

The standard postprocessor for `ConvDiff` contains `ErrorCalculator` and `OutputGenerator`. They work with the `DoFHandler`. Do they need a re-setup() after adaptive mesh refinement? Of course, applications might implement any postprocessor with additional features as a derived class. Then, however, such a call of `postprocessor->setup()` can be dangerous. So I think we need to discuss this aspect of postprocessing with AMR more fundamentally. 

My current understanding is the following: postprocessing can be anything and we don't know this here in the `Driver`. So all we can do is to prepare a member function `Postprocessor::setup_after_coarsening_and_refinement()` in which the postprocessor either re-initializes certain postprocessing functionality, or where one could assert that a certain postprocessing functionality must be in-active because it does not support AMR in its current form. If this function is virtual like `setup()`, any special application can care about this as needed. In contrast, we probably do not want to make every postprocessing functionality capable of AMR.

Reply:
Good catch! Shoudl we go for pure virtual to force implementation in some derived class, such that one is forced to think about what happens after AMR? Otherwise, this might generate bugs that are somewhat tricky to find ...
The ConvDiff applications use `OutputGenerator` and `ErrorCalculator`, yes (see exadg/include/exadg/convection_diffusion/postprocessor/postprocessor.h).

.) The `ErrorCalculator` (its `time_control`, so when it is triggered etc.) should not be reset by AMR. So I intriduced `ErrorCalculator<dim, Number>::setup_after_coarsening_and_refinement`, which just updates the DoFHandler and the Mapping.

.) Similar thing apply to `OutputGenerator`, where we can also skip output of boundary IDs etc., which is normally done at setup. Optionally, we can also include that output functionality in the setup.